### PR TITLE
Provides the types for `combine` and `combineWithAllErrors`, solves #226

### PR DESCRIPTION
**Pull request by incetarik**
Hello,

This PR solves the type problem of the `combine` and `combineWithAllErrors` methods as described [here](https://github.com/supermacro/neverthrow/issues/226).

### CODE REVIEW

**Comment by incetarik** [comment]
Just realized that there are tests depending on it, I'm now improving the types and will push another commit for each test case.

**Comment by incetarik** [comment]
Test are now completely passing, I'm working on linting now and then will push the new version.

**Comment by supermacro** [comment]
Amazing @incetarik .. currently there's a lot on my plate at work so I will aim to have provided a thorough review within a week's time - apologies for being a bit slow / unresponsive.

File: tests/typecheck-tests.ts [hunk 1]
```diff
@@ -542,7 +542,72 @@ import { ok, err, okAsync, errAsync, Result, ResultAsync } from '../src'
 
 
 (function describe(_ = 'Combine on Unbounded lists') {
-  // TODO:
-  // https://github.com/supermacro/neverthrow/issues/226
+  (function describe(_ = 'combine') {
+    (function it(_ = 'combines different results into one') {
+      type Expectation = Result<[ number, string ], Error | string[]>;
```
Comment (supermacro) [inline tests/typecheck-tests.ts:547]: This expectation is not correct - perhaps there may some miscommunication / misunderstanding of the desired goal of #226. 

The desired outcome when fixing #226 is that all the `combine` functions can correctly produce a **tuple** whose length / size is equal to the length / size of the array passed into the `combine` functions.

In other words, we need to treat the list argument to the `combine` functions as a tuple instead of a variable-length list.

Example:

When calling:

```typescript
Result.combine([
  ok(1),
  ok('string'),
  err([ 'string', 'string2' ]),
  err(new Error('error content')),
])
```

The `resultList` argument should understand that we have a fixed-length tuple whose type is:

```typescript
[ Result<number, never>, Result<string, never>, Result<never, string>, Result<never, Error> ]
```

Therefore, the `Expectation` of this test would be that we should have:

```typescript
Result<[ number, string, never, never ], Error | string[] | never>;
```

---

Additionally - in practice, `combine` would not be called with the `ok` and `err` constructors (or the `okAsync` / `errAsync` constructors), rather, we'd have a list of `Result<T, E>` values where the `T`s and the `E`s are not `unknown`.

So to make these tests a bit more realistic, I would provide explicit type arguments to the `ok` and `err` function invocations like this:

```typescript
type Expectation = Result<[ number, string, boolean ], Error | string[]>;

const result: Expectation = Result.combine([
  ok<number, string>(1),
  ok<string, string>('string'),
  err<boolean, string[]>([ 'string', 'string2' ]),
  err<boolean, Error>(new Error('error content')),
])
```

**Review by supermacro — CHANGES_REQUESTED** [review]

File: tests/typecheck-tests.ts (hunk 1 above)
Comment (incetarik) [reply tests/typecheck-tests.ts:547]: Ok as the types did already pass the tests in the examples that I have not added, I can still inspect the ones that you mention. Add your expectations and I'll see them. After this point, obviously, it will not be hard to do the rest. @supermacro

File: tests/typecheck-tests.ts [hunk 2]
```diff
@@ -542,7 +542,72 @@ import { ok, err, okAsync, errAsync, Result, ResultAsync } from '../src'
...
+        ok('string'),
+      ]);
+    });
+
+    (function it(_ = 'combines only err results into one') {
+      type Expectation = Result<never, (string | number)[]>;
+
+      const result: Expectation = Result.combineWithAllErrors([
+        err(1),
+        err('string'),
```
Comment (supermacro) [inline tests/typecheck-tests.ts:608]: :+1:

File: tests/typecheck-tests.ts [hunk 3]
```diff
@@ -542,7 +542,72 @@ import { ok, err, okAsync, errAsync, Result, ResultAsync } from '../src'
...
+
+  (function describe(_ = 'combineWithAllErrors') {
+    (function it(_ = 'combines different results into one') {
+      type Expectation = Result<[ number, string ], (Error | string[])[]>;
+
+      const result: Expectation = Result.combineWithAllErrors([
+        ok(1),
+        ok('string'),
+        err([ 'string', 'string2' ]),
+        err(new Error('error content')),
```
Comment (supermacro) [inline tests/typecheck-tests.ts:590]: Shouldn't the expectation here be:

 ```typescript
Result<[ number, string, never, never ], (Error | string[])[]>```

??

File: tests/typecheck-tests.ts [hunk 4]
```diff
@@ -542,7 +542,72 @@ import { ok, err, okAsync, errAsync, Result, ResultAsync } from '../src'
...
+      ]);
+    });
+
+    (function it(_ = 'combines only err results into one') {
+      type Expectation = Result<never, number | string>;
+
+      const result: Expectation = Result.combine([
+        err(1),
+        err('string'),
+      ]);
```
Comment (supermacro) [inline tests/typecheck-tests.ts:572]:
```suggestion
      type Expectation = Result<[ never, never ], number | string>;

      const result: Expectation = Result.combine([
        err(1),
        err('string'),
      ]);
```

**Review by supermacro — CHANGES_REQUESTED** [review]

File: src/result.ts [hunk 5]
```diff
@@ -308,3 +303,165 @@ export class Err<T, E> implements IResult<T, E> {
...
+  13,
+  14,
+  15,
+  16,
+  17,
+  18,
+  19,
+  20,
+  ...0[]
+]
```
[the remark this answers is not in the document]
Comment (incetarik) [reply src/result.ts:371]: Yes, that's right. AFAIK, TypeScript already does recursion for only fixed amount of times and it is why we have 20 there.

File: tests/typecheck-tests.ts (hunk 3 above)
Comment (incetarik) [reply tests/typecheck-tests.ts:590]: I have not considered `never` types to be in order, therefore not processed them. When I'm free, I'll go back and add those and it will work. The hardest part is already done as you see. The simplest thing would be taking this result and the parameter array (which is readonly) and until their lengths matches, adding `never`s to array is the first solution I have in my mind but as I said, I'll see this when I'm free. Should not be a big problem as we could done it until this point.

File: src/result.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (supermacro) [reply src/result.ts:371]: >  TypeScript already does recursion for only fixed amount of times

If TS already has a recursion limit, why would we then need to impose our own limit?

File: tests/typecheck-tests.ts [hunk 6]
```diff
@@ -542,7 +542,72 @@ import { ok, err, okAsync, errAsync, Result, ResultAsync } from '../src'
...
+      type Expectation = Result<[ number, string ], never[]>;
+
+      const result: Expectation = Result.combineWithAllErrors([
+        ok(1),
+        ok('string'),
+      ]);
+    });
+
+    (function it(_ = 'combines only err results into one') {
+      type Expectation = Result<never, (string | number)[]>;
```
Comment (supermacro) [inline tests/typecheck-tests.ts:604]: Just realized I missed this one - this should also be changed to `Result<never[], (string | number)[]>

**Review by supermacro — CHANGES_REQUESTED** [review]

File: src/result.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (incetarik) [reply src/result.ts:371]: TS by default tries to resolve the type recursively as much as possible and that usually ends up in hanging or simply replacing the type with `never` or `unknown` as you can expect. We just make this in that way so that the TS server will stop recursing indefinitely, i.e. not showing us "Type instantiation is excessively deep and possibly infinite" error.

**Comment by supermacro** [comment]
:warning:  **Potential Updates To ResultAsync functions and `utils.ts`**

I believe that the utility functions inside of `utils` (which are used by both the synchronous and asynchronous versions) might need to have their types updated. This might also mean that we no longer need to do typecasts such as the one [here](https://github.com/supermacro/neverthrow/pull/418/files#diff-fe54c0b0441ef7719a22666a2bb171f2db9e99c0ed52dd8fec96a9882e7c842dR37).

Also note that the async versions of `combine` and `combineWithAllErrors` is still using the `Extract*` utility types - which, in my mind, it seems like we no longer need (unless I am missing something). 

image

File: tests/typecheck-tests.ts (hunk 1 above)
Comment (incetarik) [reply tests/typecheck-tests.ts]: Hello @supermacro, I have improved the typing system and made some of the lighter for CPU, and also easier for type matching. Yet, I have not touched anything else mentioned other than this conversation page.

File: tests/typecheck-tests.ts (hunk 3 above)
Comment (incetarik) [reply tests/typecheck-tests.ts]: Updated the typing, see the commit.

File: tests/typecheck-tests.ts [hunk 7]
```diff
@@ -542,7 +542,72 @@ import { ok, err, okAsync, errAsync, Result, ResultAsync } from '../src'
...
+      type Expectation = Result<never, number | string>;
+
+      const result: Expectation = Result.combine([
+        err(1),
+        err('string'),
+      ]);
+    });
+
+    (function it(_ = 'combines empty list results into one') {
+      type Expectation = Result<never, never>;
```
[the remark this answers is not in the document]
Comment (incetarik) [reply tests/typecheck-tests.ts]: Done

File: tests/typecheck-tests.ts (hunk 4 above)
Comment (incetarik) [reply tests/typecheck-tests.ts]: Done

File: tests/typecheck-tests.ts (hunk 6 above)
Comment (incetarik) [reply tests/typecheck-tests.ts]: @supermacro Could you please see the changes? Now I have updated the types as well, they are now more precise and in-order as well.

**Comment by incetarik** [comment]
> ⚠️ **Potential Updates To ResultAsync functions and `utils.ts`**
> 
> I believe that the utility functions inside of `utils` (which are used by both the synchronous and asynchronous versions) might need to have their types updated. This might also mean that we no longer need to do typecasts such as the one [here](https://github.com/supermacro/neverthrow/pull/418/files#diff-fe54c0b0441ef7719a22666a2bb171f2db9e99c0ed52dd8fec96a9882e7c842dR37).
> 
> Also note that the async versions of `combine` and `combineWithAllErrors` is still using the `Extract*` utility types - which, in my mind, it seems like we no longer need (unless I am missing something).
> 
> image

I have done nothing for this. I may see this after you see the commits I have sent and approve, I think this is out of the current conversation but also something that could be done as well, right.

File: src/result.ts [hunk 8]
```diff
@@ -308,3 +315,171 @@ export class Err<T, E> implements IResult<T, E> {
...
+      >
+    : never // Impossible
+  : Collected
+
+// Transposes an array
+//
+// A          - The array source
+// Transposed - The collected transposed array
+// Depth      - The maximum depth.
+type Transpose<A, Transposed extends unknown[][] = [], Depth extends number = 10> = A extends [
```
Comment (supermacro) [inline src/result.ts:379]: Could we export this to have 1 or 2 type-level tests written for it, please?

File: tests/typecheck-tests.ts [hunk 9]
```diff
@@ -551,6 +552,71 @@ import { ok, err, okAsync, errAsync, fromSafePromise, Result, ResultAsync } from
...
+      type Expectation = Result<[ number, string ], never[]>;
+
+      const result: Expectation = Result.combineWithAllErrors([
+        ok(1),
+        ok('string'),
+      ]);
+    });
+
+    (function it(_ = 'combines only err results into one') {
+      type Expectation = Result<never, [number, string]>;
```
Comment (supermacro) [inline tests/typecheck-tests.ts:614]:
```suggestion
      type Expectation = Result<never[], [number, string]>;
```

This is odd, it seems like there is different behaviour when passing similar arguments. Compare this expectation to this one:

https://github.com/supermacro/neverthrow/pull/418/files#diff-5e6df76f1ff5c2ddfbb0bee2e515236794778a429792a91d3b853bfecc884c16R577

**Review by supermacro — CHANGES_REQUESTED** [review]

**Comment by supermacro** [comment]
> I have done nothing for this. I may see this after you see the commits I have sent and approve, I think this is out of the current conversation but also something that could be done as well, right.

I'm happy with your approach to the async side, but we also need to apply this work to the async side:

image

the `test` variable is currently a `ResultAsync<(string | number | Date)[], string | boolean | Error>`, but it should instead be `ResultAsync<[ number, string, Date ], string | boolean | Error>` based on the work you've done in this PR.

File: tests/typecheck-tests.ts (hunk 9 above)
Comment (incetarik) [reply tests/typecheck-tests.ts:614]: @supermacro  These two are different, one is a tuple where the other is an array. The tuple has fixed size where the array does not. I have preferred to use tuple as it would be providing a better type and length info.

File: src/result.ts (hunk 8 above)
Comment (incetarik) [reply src/result.ts:379]: @supermacro

**Comment by incetarik** [comment]
> > I have done nothing for this. I may see this after you see the commits I have sent and approve, I think this is out of the current conversation but also something that could be done as well, right.
> 
> I'm happy with your approach to the synchronous side, but we also need to apply this work to the async side:
> 
> image
> 
> the `test` variable is currently a `ResultAsync<(string | number | Date)[], string | boolean | Error>`, but it should instead be `ResultAsync<[ number, string, Date ], string | boolean | Error>` based on the work you've done in this PR.

I have not done anything in the async side. I do not expect it to be anything hard as well. Yet, I have never considered them as the problem/talk was for combine and its other only, not async.

**Comment by incetarik** [comment]
Good news @supermacro! The typing is also added for `ResultAsync` and it is precise as `Result`. After you see them, I have left the removals of the unused/unnecessary types to you as you may change them or reorder/update etc. Have a nice day!

**Comment by supermacro** [comment]
Amazing! Thank you. I don't know if I'll have time to do a final read through of everything today, but I will try to do so by end of the week. Otherwise, I will definitely do a review by next week at the latest. And then I can send over that bounty :slightly_smiling_face:  

From a quick skim, I did notice that you added overload type definitions to the `ResultAsync.combine` and `ResultAsync.combineWithAllErrors` methods - it would be helpful to get some context on why that was done and what benefit it has.

**Comment by incetarik** [comment]
> Amazing! Thank you. I don't know if I'll have time to do a final read through of everything today, but I will try to do so by end of the week. Otherwise, I will definitely do a review by next week at the latest. And then I can send over that bounty 🙂
> 
> From a quick skim, I did notice that you added overload type definitions to the `ResultAsync.combine` and `ResultAsync.combineWithAllErrors` methods - it would be helpful to get some context on why that was done and what benefit it has.

@supermacro Of course, it has a simple explanation. The TypeScript will always try to match the overloads from top to bottom. Whenever the match occurs the others are skipped. This becomes pretty useful when you want to handle the parameter for the cases where one time you pass something has narrowed, specific type such as `[1, 2, 3]` instead of `number[]`. Here you may want to do something else when you have a more accurate type. I use this to match for each element in the array. Secondly, if user passes something more generic than that, then I handle it differently and return some more general type as the return value. This is why we have overloads for these types. If you look closely, the types actually changing nothing in terms of types but the number of elements. If we are sure that there is at least one element in the array passed to the function so we have better constraints to take advantage of, otherwise we fallback to the default implementation.

File: tests/typecheck-tests.ts [hunk 10]
```diff
@@ -551,6 +561,126 @@ import { ok, err, okAsync, errAsync, fromSafePromise, Result, ResultAsync } from
...
+      type Expectation = Result<[ number, string ], never>;
+
+      const result: Expectation = Result.combine([
+        ok(1),
+        ok('string'),
+      ]);
+    });
+
+    (function it(_ = 'combines only err results into one') {
+      type Expectation = Result<[ never, never ], number | string>;
```
Comment (supermacro) [inline tests/typecheck-tests.ts:586]: ### Edge case still not working as expected

When a list of `err` values is passed into any variation of `combine` (whether it's synchronous or asynchronous) **AND** there is no type argument for the success type `T` passed in (thus, the `T` becomes `never`), then we would expect to have a type akin to the one that you have written here on line 586.

**However**, when I try the code out on my machine, the inferred type does not match this expectation:

image

The Expectation is correct in this file, but for some reason the type inference logic is not returning what we expect - and for some reason the code still compiles (I don't understand how `never` can be assigned to `[ never, never ]`).

**Review by supermacro — CHANGES_REQUESTED** [review]

**Comment by supermacro** [comment]
[This](https://github.com/supermacro/neverthrow/pull/418#discussion_r996471068) is the last thing remaining! I'm super excited to get this merged. Great job @incetarik

**Comment by incetarik** [comment]
> 

@supermacro Thank you! When I'm free, I'll see this and I think it will be easy to solve and I'll be notifying you. I'm expecting this to be fixed in a day.

File: tests/typecheck-tests.ts (hunk 10 above)
Comment (incetarik) [reply tests/typecheck-tests.ts:586]: @supermacro I have fixed this pretty easily actually, see the line `462` and `432` in `result.ts` file.

I have done this only for the case you have mentioned. If you want to apply this logic to other places as well, simply change `EmptyArrayToNever<...>` to `EmptyArrayToNever<..., 1>`. This extra `1` argument will cause to keep `[never, ...never[]]`-like array as it is. Add `1` here if you want to get precise type such as `[never, never, never]` instead of `never` which was the previous default that would be occurring if all of the elements of the array is `never`.

**Review by supermacro — APPROVED** [review]

**Comment by supermacro** [comment]
I will be doing a review of documentation and then creating the release today / tomorrow. Thank you for your contribution @incetarik ... sending over bounty payment today :slightly_smiling_face:
